pkg/kn/commands/completion_helper.go

```diff
@@ -0,0 +1,129 @@
+// Copyright © 2021 The Knative Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package commands
+
+import (
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+type completionConfig struct {
+	params     *KnParams
+	command    *cobra.Command
+	args       []string
+	toComplete string
+	target     string
+}
+
+var (
+	resourceToFuncMap = map[string]func(config *completionConfig) ([]string, cobra.ShellCompDirective){
+		"service": completeService,
+	}
+)
+
+// ResourceNameCompletionFunc will return a function that will autocomplete the name of
+// the resource based on the subcommand
+func ResourceNameCompletionFunc(p *KnParams) func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
```

Comment:
Is there a case when *not* `cobra.ShellCompDirectiveNoFileComp` is returned ? Because then you could remove that return value.

Reply:
`cobra.ShellCompDirectiveNoFileComp` is being returned in all cases. The function being returned needs to be of this specific signature to be used as a `ValidArgsFunc`